### DISCUSSION TITLE
is there any way to toogle case?

### DISCUSSION THREAD

**Discussion by ousia** [orbitalquark/textadept#213] [Q&A] [answered]
Is there any way to toggle case in selection (upper/lower)?

I would like also to assign a single keystroke to that.

Many thanks for your help.

**Comment by rgieseke** [comment]
You mean a single shortcut to toggle the current state?

Otherwise there is this:

> Convert the selected text to upper or lower case via Ctrl+Alt+U or Ctrl+Alt+Shift+U, respectively, on Windows, Linux, and BSD; ^U or ^⇧U, respectively, on macOS; and M-^U or M-^L in the terminal version.

**Reply by ousia** [reply]
I need in a single shortcut.

I think this is the conditional, but I cannot get it working:

``` lua
keys['ctrl+alt+u'] = function()
  if Upper_Case then
    buffer.lower_case()
    Upper_Case = false
  else
    buffer.lower_case()
    Upper_Case = true
  end
end
```

I get only uppercase (even with `buffer.lower_case()`).

**Reply by rgieseke** [reply]
I think you need to have your state variable definition outside of your function:

```lua
local upper_cased
keys['ctrl+alt-u'] = function()
  if upper_cased then
    buffer.lower_case()
    upper_cased = false
  else
    buffer.upper_case()
    upper_cased = true
  end
end
```

**Reply by ousia** [reply]
You’re right, but I cannot mark your reply as an answer ☹️.

Sorry, the people who designed this haven’t considered that a reply could contain the answer.

**Reply by rgieseke** [reply]
No worries, i find the interface with the highlighted reply confusing anyways ...

**Comment by ousia** [comment, marked as answer]
@rgieseke, this works fine:

``` lua
keys['ctrl+alt+u'] = function()
  if Upper_Case then
    buffer.lower_case()
    Upper_Case = false
  else
    buffer.upper_case()
    Upper_Case = true
  end
end
```

I realized I was modifying a backup `init.lua` (so my approach couldn’t work at all 😅).

It seems that for Lua if _x_ is nil, _x_ is both not true and not false.

``` lua
if UpperCase then print("yes") else print("no") end

if UpperCase == true then print("yes") else print("no") end

if UpperCase == false then print("yes") else print("no") end

if not UpperCase then print("yes") else print("no") end
```

The first three conditionals print _no_, the fourth one prints _yes_.

This did the trick here.

**Reply by rgieseke** [reply]
Excellent! Btw. you can also use `ui.statusbar_text = 'Upper: '..UpperCase` for debugging/developing ...

**Reply by rgieseke** [reply]
That would have to be `ui.statusbar_text = 'Upper: '..tostring(UpperCase)`

**Reply by ousia** [reply]
Good to know that, @rgieseke.
